design/helm.md

```diff
@@ -0,0 +1,145 @@
+
+
+## Goals
+- Allow any Helm chart to be specified as part of a Kurl spec.
+
+## Non Goals
+
+- Verify authenticity, trustworthyness and "goodness" of helm addons.
+- Validate Helm values.
+
+## Background
+
+Users of Kurl have a limited set of available cluster addons and often want to include ones that are unsupported.
+There is friction in the [kurl development process](https://github.com/replicatedhq/kURL/blob/master/test/dev.md), [contributing new addons](https://kurl.sh/docs/add-on-author/) and addon maintanence. 
+
+Helm is a popular tool for managing deployment configuration that allows application/addon developers or the community to maintain a consistent interface for deployment through a chart.
+By allowing Kurl users to specifify a chart and its values, we can allow customization of kutl with new addons while delegating addons support and maintance to its community.
+
+## High-Level Design
+
+Kurl will bundle the Helm binary into the common tools bundle already delivering krew and kubectl plugins. The install.sh script will perform the following in the online case:
+1. Unarchive Helm and move into the `$PATH` and `/usr/bin`
+1. For each release, check if the release exists. 
+1. If the release exists, perform a `helm upgrade`.
+1. If the release doesn't exist, perform a `helm install`
+
+For airgapped the chart and artifacts need to be packaged into a bundle. The builder should perform the following:
+1. `helm pull` the repo
+1. `helm template` to render the chart with the provided values
+1. Generate an adhoc `manifest` file by parsing `image` specs from the rendered yaml.
+1. Generate a bundle including the chart + containers
+
+In the airgapped case, install.sh will perform the following:
+1. Unpack the helm bundle on initialization
+1. Load docker containers as part of the initialization
+1. Unarchive the chart and install with the installer spec values during the install phase
+
+## Detailed Design
+
+The yaml design is informed by the [Flux Helm Operator CRD](https://github.com/fluxcd/helm-operator/blob/master/docs/references/helmrelease-custom-resource.md). 
+Another spec is [Rancher's Helm CRD](/var/lib/rancher/rke2/server/manifests/) that works with k3s and RKE2.
+
+The proposal is to implement option1 in the updated Kurl Installer Spec:
+```yaml
+apiVersion: "cluster.kurl.sh/v1beta1"
+kind: "Installer"
+metadata: 
+  name: "latest"
+spec: 
+  kubernetes: 
+    version: "latest"
+  weave: 
+    version: "latest"
+  containerd: 
+    version: "latest"
+  helm:                 # Optional Future Capability
+    version: "latest"
+  helmReleases:
+    # Basic Idea, Option 1
+    - release: nginx-ingress
+      namespace: kube-system    # Optional
+      chart: 
+        name: bitnami/nginx
+        repo: https://charts.bitnami.com/bitnami 
+        version: 8.5.1          # Optional - uses latest if not provided
```

Comment:
Using latest could be problematic, particularly with airgap since airgap bundle would be created at a certain point in time. Could/would we detect upstream changes and create new airgap bundles proactively?

Reply:
I agree it's problematic but we could give consumers the option the same as other kurl addons to be consistent. We'll have to hash airgap bundles anyway since any values change could generate a new bundle. The hash + the helm manifest will help us identify the installed version.

As for proactively building bundles, I imagine it's on demand now and it would be a big change to monitor upstream containers/images, but it could be done. I don't really know how the bundling system works yet TBH.

Reply:
Yeah, I forgot we already have that problem with airgap bundles and `latest`.